scripts/pre_commit_linter.py

```diff
@@ -202,6 +202,32 @@
         'excluded_files': (),
         'excluded_dirs': ()
     },
+    {
+        'regexp': re.compile(r'toThrow\('),
+        'message': 'Please use \'toThrowError\' instead of '
+                   '\'toThrow\'',
+        'excluded_files': (),
+        'excluded_dirs': ('extensions/interactions/LogicProof/static/js')
```

Comment:
Why is this file excluded?

Reply:
there are many places in that folder `throw error (as a variable)`. I wasn't sure to I change it or not, so asked here but no one replied so, I excluded the directory. Should I change them?

Reply:
@seanlip second thoughts?

Reply:
and the 

> > ```
> >    }).toThrow(
> >       sharedErrorWrapper(
> >         'This line uses q, so you need to have an earlier line proving ' +
> >         'that q is true.', 0, 'missing_antecedent', 'logic'));
> > ```
> > 
> > 
> > ```
> > .toThrow(
> >       {
> >         message: (
> >           'The name \'we\' is reserved for vocabulary and so cannot ' +
> >           'be used here.')
> >       }
> >     );
> > ```
> 
> @kevinlee12 how do i handle these . My regular expressions are unable to handle these kind of errors
> 
> > ```
> >   {
> >         'regexp': re.compile(r'\btoThrow\('),
> >         'message': 'Please use \'toThrowError\' instead of '
> >                    '\'toThrow\'',
> >         'excluded_files': (),
> >         'excluded_dirs': ()
> > ```

Reply:
Yes, please change it. Having lots of cases is not a sufficient reason to exclude a file/folder from a lint check.

Reply:
@seanlip I was suggested to standardized the `toThrow` to `toThrowError`
according to that should I change the `toThrow(sharedErrorWrapper` to `toThrowError`
and `toThrow({message:(message)}` to `toThrowError`
OR 
should I modified the regex so that it won't catch any error if it used like that`toThrow(sharedErroWrapper`

Reply:
Hm, looks like we use Custom errors in those files (and there does not seem to be a way to capture this using regex alone). However, could we maybe enforce that all custom errors should end with Error? @abhayraghuwanshi would this solve the problem for some cases?

Reply:
HI @kevintab95 
and I thought a way to handle the custom error
for example, if we find `toThrowError(something(` using regex we would suggest them to use 
`toThrowError(somethingCustomError` instead

if we change the `node_modules/@types/jasmine/ts3.1/index.d.ts`
```
toThrow(expected?: any): boolean;
toThrowError(message?: string | RegExp): boolean;
``` 
here the `toThrow` is expecting `type:any` but `toThrowError` we can only throw regex or string. and changing the type to any will solve the problem.

So, Do we need to standardized the `toThow` to `toThrowError`  because we cannot handle the objects in this case.

Reply:
Er.. I don't think we should be making changes to `node_modules/@types/jasmine/ts3.1/index.d.ts`. Can you give an example of what would fail for the `toThrowError(..` stuff with your current changes? Thanks!

Reply:
when we use` toThrowError(sharedWrapperError` in the case of `toThrow(sharedWrapperError` gives the following error
the shared wrapper function is returning the object
````
    var sharedErrorWrapper = function(message, line, code, category) {
        return {
          message: message,
          line: line,
          code: code,
          category: category
        };
      };
````
````
No overload matches this call.
  The last overload gave the following error.
    Argument of type '{ message: any; line: any; code: any; category: any; }' is not assignable to parameter of type 'new (...args: any[]) => Error'.
      Type '{ message: any; line: any; code: any; category: any; }' provides no match for the signature 'new (...args: any[]): Error'.ts(2769)
index.d.ts(591, 9): The last overload is declared here.
````

Reply:
Oh, I see. Thanks @abhayraghuwanshi for pointing that out! I just noticed that there is a TODO in [shared.ts](https://github.com/oppia/oppia/blob/develop/extensions/interactions/LogicProof/static/js/shared.ts#L48) that says we must use Error objects instead of custom objects. In that case, I think it is fine to handle this separately (ie. once the TODO stuff is done, we should enable the linter to lint these files). @seanlip, @kevinlee12  what are your thoughts?

Reply:
Please fix the TODO as part of this PR. (Note that the TODO seems to be referring to the code above it, not the code below it.)

Reply:
@seanlip I'll fix it after 6th, as my exams are going on 
thanks

Reply:
@seanlip  Is this the thing we are talking in the Todo 
https://stackoverflow.com/questions/783818/how-do-i-create-a-custom-error-in-javascript
thanks

Reply:
Yes, something along those lines, but read all the SO answers carefully to get a sense for which approach is the best to adopt.

Reply:
Also, make sure to test the interaction manually after your changes to ensure that everything is still working correctly and that answers can be submitted and evaluated.

Reply:
hii @seanlip I have tried to do the Todo but I am not sure how to proceed further( as the todo says to make the custom error prototype of an Error object) 
````
  var UserError = function(code, parameters) {
    this.name = 'UserError';
    this.code = code;
    this.parameters = parameters || {};
  };
  UserError.prototype = Error.prototype;
  UserError.prototype.name = 'UserError';

  // These errors already have messages which are user-readable.
  var PreRenderedUserError = function(messages, code) {
    this.name = 'PreRenderedUserError';
    this.messages = messages;
    this.code = code;
  };
  PreRenderedUserError.prototype = Error.prototype;
  PreRenderedUserError.prototype.name = 'PreRenderedUserError';
````
should they return the string like an Error.
Could you please clearify it a little more
Thanks

Reply:
Hi @abhayraghuwanshi, thank you for following up on this!

I had a chat with the author of the logic proof interaction for what to do here. We decided that we are going to look into deprecating it, since it is not space-efficient, and we are running into maintenance issues such as the one you are describing here.

Given this decision, I think it is fine to exclude the file for now. Please add a clear comment above the line in the pre-commit linter file to explain why we are doing this, so that other developers know that it is a very special case and that they should not add more things there as well.

Sorry for the back-and-forth on this, but I do want to clarify that the technical information you provided was necessary in order to make this decision! I hope this resolves the conversation, but let me know if you have more questions. Thanks!

Reply:
![Screenshot from 2020-03-11 11-54-32](https://user-images.githubusercontent.com/45534131/76436987-478b3500-638f-11ea-9007-d8d00697b568.png)
@seanlip  there is one more problem with the linter in was told to standardized the 
`throw err;` to `throw new Error(err);` as seen above when the err is an object the error will be thrown like the `Uncaught Error[object object]` something like above 
what should be do when this type of senario arrises

Reply:
If we are ignoring this file, then does this case occur (where the error is not a string)?  If so, where does it occur?

Reply:
https://github.com/oppia/oppia/pull/8522/files#diff-f212e0f6c92196b87df08e0e8e460b1d
https://github.com/oppia/oppia/pull/8522/files#diff-b6e084d7c98d0c8637b0bc5999ef2624
https://github.com/oppia/oppia/pull/8522/files#diff-d8160714a691a346dc4594acdab35777
https://github.com/oppia/oppia/pull/8522/files#diff-fbb9a6ab7eab26c37730e2617360ddc6
these are the list of changes I did in the where the throw err is an object and although I have excluded the logicProof  from the linter but maybe this thing create problem in future 
![Screenshot from 2020-03-12 01-33-04](https://user-images.githubusercontent.com/45534131/76491072-282def80-6403-11ea-9421-e4f23c3d3101.png)

Reply:
Ah, OK. If it is straightforward to change these to strings, and nothing else in the codebase uses the error raised from here and treats it as an object, then it is fine to change these to strings. Does that solve the problem?

Reply:
@seanlip i guess this will solve the problem .
If i have any doubts again I'll let you know
Thanks

Reply:
@seanlip i have updated them like 
````
     var additionalInfo = (
          'name: ' + err.name +
          '\nmessage: ' + err.message +
          '\nstack: ' + err.stack
        );
        throw new Error(additionalInfo);
````
PTAL
thanks